### PR TITLE
feat(oidc): support generic OIDC providers

### PR DESCRIPTION
**Pull request by GeorgeMac**
This is a PR based on the `gm/oidc-google` branch. I thought I would propose it here first.
If we like it, we can merge it into that and then merge that into `main`.

This makes the whole OIDC package generic for multiple providers.
The `providers` key is simply a map of any user-defined provider name to provider config.

This leads to dynamic OIDC endpoints. e.g.

```yaml
authentication:
  methods:
    oidc:
      providers:
        foo:
          // ...
```

This will lead to the following endpoints:

- /auth/v1/method/oidc/foo/authorize
- /auth/v1/method/oidc/foo/callback

I did a run through with Google, and it works as normal.

### CODE REVIEW

File: internal/config/authentication.go [hunk 1]
```diff
@@ -148,22 +149,18 @@ type AuthenticationMethodTokenConfig struct {
...
-
-// AuthenticationMethodOIDCProviders contains a set of providers for the OIDC authentication method.
-type AuthenticationMethodOIDCProviders struct {
-	Google *AuthenticationMethodOIDCProviderGoogle `json:"google,omitempty" mapstructure:"google"`
+	Enabled   bool                                        `json:"enabled,omitempty" mapstructure:"enabled"`
+	Providers map[string]AuthenticationMethodOIDCProvider `json:"providers,omitempty" mapstructure:"providers"`
+	Cleanup   *AuthenticationCleanupSchedule              `json:"cleanup,omitempty" mapstructure:"cleanup"`
 }
 
 // AuthenticationOIDCProviderGoogle configures the Google OIDC provider credentials
```
Comment (markphelps) [inline internal/config/authentication.go:157]:
```suggestion
// AuthenticationOIDCProvider configures provider credentials
```

File: internal/config/config.go [hunk 2]
```diff
@@ -209,3 +221,44 @@ func stringToSliceHookFunc() mapstructure.DecodeHookFunc {
...
 }
+
+// trie is a prefix trie used to create a map of potential
+// key paths from the current environment.
+// This is used to support pre-emptively binding env vars
+// for the keys of map[string]T which we don't know aheaad of
+// time.
+type trie map[string]trie
+
+func (t trie) getPrefix(parts ...string) (trie, bool) {
```
Comment (markphelps) [inline internal/config/config.go:232]: could we add some more tests around this method and the function below it if possible?

**Review by markphelps — COMMENTED** [review]
makes sense to me! looks good!! 

two minor comments

File: internal/config/config.go (hunk 2 above)
Comment (GeorgeMac) [reply internal/config/config.go]: I've taken another pass in a slightly different way. Which also allowed me to test it more thoroughly.
It also accounts for something this implementation couldn't do. Which was keys in maps with `_` in them.
See what you think.

**Review by markphelps — APPROVED** [review]
🕺

**Pull request merged**